### PR TITLE
Initial implementation of dataplane + interface config handler

### PR DESCRIPTION
**Pull request by DanG100**

### CODE REVIEW

File: dataplane/handlers/interface.go [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+// reconcile compares the interface config with state and modifies state to match config.
+func (ni *Interface) reconcile(config *oc.Interface) {
+	ni.mu.RLock()
+	defer ni.mu.RUnlock()
+
+	tapName := engine.IntfNameToTapName(config.GetName())
+	state := ni.state[config.GetName()]
+
+	if config.GetOrCreateEthernet().MacAddress != nil {
```
Comment (wenovus) [inline dataplane/handlers/interface.go:139]: What if mac-address is nil? Should the TAP interface be deleted? And then re-created? Is it possible to just reset the HW address and it auto-disables?

File: dataplane/handlers/interface.go [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+		if config.GetEthernet().GetMacAddress() != state.GetEthernet().GetMacAddress() {
+			log.Infof("setting interface %s hw-addr %q", engine.IntfNameToTapName(config.GetName()), config.GetEthernet().GetMacAddress())
+			if err := kernel.SetInterfaceHWAddr(engine.IntfNameToTapName(config.GetName()), config.GetEthernet().GetMacAddress()); err != nil {
+				log.Warningf("Failed to set mac address of port: %v", err)
+				return
+			}
+		}
+	}
+	if config.GetOrCreateSubinterface(0).Enabled != nil {
```
Comment (wenovus) [inline dataplane/handlers/interface.go:148]: What if mac-address is nil? Should the TAP interface be deleted?

File: dataplane/handlers/interface.go [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+		configPL := addr.PrefixLength
+		var stateIP *string
+		var statePL *uint8
+		stateAddr := state.GetSubinterface(0).GetIpv4().GetAddress(addr.GetIp())
+		if stateAddr != nil {
+			stateIP = stateAddr.Ip
+			statePL = stateAddr.PrefixLength
+		}
+
+		if configIP != nil && configPL != nil && (stateIP == nil || *statePL != *configPL) {
```
Comment (wenovus) [inline dataplane/handlers/interface.go:169]: what if the stateIP doesn't match configIP but the PLs match?

File: dataplane/handlers/interface.go [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+// Stop stops all watchers.
+func (ni *Interface) Stop() {
+	ni.watchCancelFn()
+	close(ni.linkDoneCh)
+	close(ni.addrDoneCh)
+}
+
+// reconcile compares the interface config with state and modifies state to match config.
+func (ni *Interface) reconcile(config *oc.Interface) {
```
Comment (wenovus) [inline dataplane/handlers/interface.go:132]: (no action required) I like this pattern: it cleanly separates out the reconciliation action from the watch action.

File: dataplane/handlers/interface.go [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+		}
+
+		if configIP != nil && configPL != nil && (stateIP == nil || *statePL != *configPL) {
+			log.Infof("Config IP: %v, Config PL: %v. State IP: %v, State PL: %v", addr.GetIp(), addr.GetPrefixLength(), stateAddr.GetIp(), stateAddr.GetPrefixLength())
+			log.Infof("setting interface %s ip %s/%d", tapName, *configIP, *configPL)
+			if err := kernel.SetInterfaceIP(engine.IntfNameToTapName(config.GetName()), *configIP, int(*configPL)); err != nil {
+				log.Warningf("Failed to set ip address of port: %v", err)
+				return
+			}
+		}
```
Comment (wenovus) [inline dataplane/handlers/interface.go:196]: what if the stateIP doesn't match configIP but the PLs match?

File: dataplane/internal/kernel/kernel.go [hunk 6]
```diff
@@ -24,37 +24,50 @@ import (
...
+	ipNet.IP = net.ParseIP(ip)
+	if ipNet.IP == nil {
+		return fmt.Errorf("failed to parse ip")
 	}
-	for i := range currIPs {
-		if err := netlink.AddrDel(link, &currIPs[i]); err != nil {
-			return fmt.Errorf("failed to add delete existing IP: %w", err)
-		}
+	ipNet.Mask = net.CIDRMask(prefixLen, 128)
+	if ipNet.IP.To16().To4() != nil {
```
Comment (wenovus) [inline dataplane/internal/kernel/kernel.go:54]: what is this pattern? i'm not understanding it.

File: dataplane/internal/kernel/kernel.go [hunk 7]
```diff
@@ -24,37 +24,50 @@ import (
...
+		ipNet.Mask = net.CIDRMask(prefixLen, 32)
+	}
+	if err := netlink.AddrReplace(link, &netlink.Addr{IPNet: ipNet}); err != nil {
+		return fmt.Errorf("failed to add ip to link: %w", err)
+	}
+	return nil
+}
+
+// SetInterfaceIP sets the IP addresses of a network interface.
+func DeleteInterfaceIP(name string, ip *net.IPNet) error {
```
Comment (wenovus) [inline dataplane/internal/kernel/kernel.go:64]: update comment

File: dataplane/handlers/interface.go [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+	}()
+
+	return nil
+}
+
+// Stop stops all watchers.
+func (ni *Interface) Stop() {
+	ni.watchCancelFn()
+	close(ni.linkDoneCh)
+	close(ni.addrDoneCh)
```
Comment (wenovus) [inline dataplane/handlers/interface.go:128]: nit: have a internal state for whether has started to guard this so this never panics.

File: dataplane/handlers/interface.go [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+			log.Infof("Config IP: %v, Config PL: %v. State IP: %v, State PL: %v", addr.GetIp(), addr.GetPrefixLength(), stateAddr.GetIp(), stateAddr.GetPrefixLength())
+			log.Infof("setting interface %s ip %s/%d", tapName, *configIP, *configPL)
+			if err := kernel.SetInterfaceIP(engine.IntfNameToTapName(config.GetName()), *configIP, int(*configPL)); err != nil {
+				log.Warningf("Failed to set ip address of port: %v", err)
+				return
+			}
+		}
+	}
+	// TODO: delete IPs
+}
```
Comment (wenovus) [inline dataplane/handlers/interface.go:199]: After reconciliation, need to update state data to the datastore

File: dataplane/server.go [hunk 10]
```diff
@@ -20,79 +20,93 @@ import (
...
-	if req.Hwaddr != "" {
-		addr, err := net.ParseMAC(req.Hwaddr)
-		if err != nil {
-			return nil, status.Errorf(codes.InvalidArgument, "invalid MAC address %q: parse err %v", req.Hwaddr, err)
-		}
-		if err := setInterfaceHWAddr(tapName, addr); err != nil {
-			return nil, status.Errorf(codes.Internal, "failed to set HW addr: %v", err)
-		}
+
+	list, err := net.Listen("tcp", "127.0.0.1:0")
```
[the remark this answers is not in the document]
Comment (DanG100) [reply dataplane/server.go:58]: port 0 mean pick a random open port

File: dataplane/handlers/interface.go [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+	iface.GetOrCreateEthernet().MacAddress = ygot.String(lu.Attrs().HardwareAddr.String())
+	iface.Ifindex = ygot.Uint32(uint32(lu.Attrs().Index))
+	iface.Enabled = ygot.Bool(lu.Attrs().Flags == net.FlagUp)
+	var operStatus oc.E_Interface_OperStatus
+	switch lu.Attrs().OperState {
+	case netlink.OperDown:
+		operStatus = oc.Interface_OperStatus_DOWN
+	case netlink.OperUp:
+		operStatus = oc.Interface_OperStatus_UP
+	}
```
[the remark this answers is not in the document]
Comment (DanG100) [reply dataplane/handlers/interface.go:231]: the zero value is UNSET, so a default case isn't needed

File: dataplane/handlers/interface.go (hunk 2 above)
Comment (DanG100) [reply dataplane/handlers/interface.go:148]: why would remove an interface?

File: dataplane/handlers/interface.go (hunk 1 above)
Comment (DanG100) [reply dataplane/handlers/interface.go:139]: deleting the mac address should reset it the default. probably not needed yet added todo

File: dataplane/handlers/interface.go (hunk 3 above)
Comment (DanG100) [reply dataplane/handlers/interface.go:169]: that handled by deleting ip that exist in state but not config for which there's a todo

File: dataplane/handlers/interface.go (hunk 5 above)
Comment (DanG100) [reply dataplane/handlers/interface.go:196]: see earlier comment

File: dataplane/internal/kernel/kernel.go (hunk 6 above)
Comment (DanG100) [reply dataplane/internal/kernel/kernel.go:54]: added comment

File: dataplane/handlers/interface.go (hunk 9 above)
Comment (DanG100) [reply dataplane/handlers/interface.go:199]: added todo

File: lemming.go [hunk 12]
```diff
@@ -116,6 +116,18 @@ func New(lis net.Listener, targetName string, opts ...grpc.ServerOption) (*Devic
...
+	if err := dplane.Start(context.Background()); err != nil {
+		return nil, err
+	}
+	hal, err := dplane.HALClient()
+	if err != nil {
+		return nil, err
+	}
+
+	startSysrib(&sysrib.Dataplane{HALClient: hal})
+
```
[the remark this answers is not in the document]
Comment (DanG100) [reply lemming.go]: fixed init order

File: dataplane/internal/kernel/kernel.go [hunk 13]
```diff
@@ -24,37 +24,50 @@ import (
...
+	ipNet.IP = net.ParseIP(ip)
+	if ipNet.IP == nil {
+		return fmt.Errorf("failed to parse ip")
 	}
-	for i := range currIPs {
-		if err := netlink.AddrDel(link, &currIPs[i]); err != nil {
-			return fmt.Errorf("failed to add delete existing IP: %w", err)
-		}
+	ipNet.Mask = net.CIDRMask(prefixLen, 128)
+	if ipNet.IP.To16().To4() != nil { // If ip is IPv4.
```
Comment (wenovus) [inline dataplane/internal/kernel/kernel.go:54]: Why is To16() needed?

**Review by wenovus — APPROVED** [review]

File: dataplane/internal/kernel/kernel.go (hunk 13 above)
Comment (DanG100) [reply dataplane/internal/kernel/kernel.go]: good catch, removed

**Pull request merged**